File: android/app/src/main/java/io/atlas/servidor/core/SessionWatcher.kt
package io.atlas.servidor.core

import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Vigia de inatividade — segurança do servidor logado.
 *
 * Depois de [TIMEOUT_MS] (10 min) sem NENHUMA interação do usuário, marca a sessão como
 * expirada. A UI (AtlasNavHost) observa [expired], faz logout e volta pro login com o
 * aviso "Sua sessão expirou".
 *
 * O timer só corre enquanto [arm]ado (usuário numa área autenticada). Cada toque na tela
 * chama [touch] (via MainActivity.onUserInteraction) e adia o prazo. O timer também dispara
 * com o app em segundo plano — se o servidor voltar depois de 10 min parado, cai no login.
 */
object SessionWatcher {
    private const val TIMEOUT_MS = 10L * 60L * 1000L // 10 minutos

    private val handler = Handler(Looper.getMainLooper())
    private val expireRunnable = Runnable { if (armed) expired = true }
    private var armed = false

    /** Sessão expirou por inatividade? A UI observa e reage (logout + login). */
    var expired by mutableStateOf(false)
        private set

    /** Liga o vigia — chamar quando o usuário entra numa área logada. */
    fun arm() {
        armed = true
        expired = false
        reschedule()
    }

    /** Desliga o vigia — logout manual/expiração já tratados; não faz sentido continuar. */
    fun disarm() {
        armed = false
        handler.removeCallbacks(expireRunnable)
    }

    /** Registra interação do usuário — adia o timeout. No-op se não estiver armado. */
    fun touch() {
        if (armed) reschedule()
    }

    /** Consumido pela tela de login depois de exibir o aviso de expiração. */
    fun consumirExpiracao() {
        expired = false
    }

    private fun reschedule() {
        handler.removeCallbacks(expireRunnable)
        handler.postDelayed(expireRunnable, TIMEOUT_MS)
    }
}

File: ios/AtlasServidor/Core/SessionWatcher.swift
import SwiftUI
import Combine

/// Vigia de inatividade — segurança do servidor logado.
/// Espelha `core/SessionWatcher.kt` do Android: 10 minutos sem interação →
/// logout automático e volta pro login com "Sua sessão expirou".
///
/// No iOS não existe `onUserInteraction`, então a detecção de toque é feita por
/// um gesto simultâneo instalado na raiz da navegação (ver `.detectaInteracao()`).
/// O tempo em segundo plano também conta: ao voltar, comparamos o instante da
/// última interação com o relógio.
@MainActor
final class SessionWatcher: ObservableObject {
    static let shared = SessionWatcher()
    private init() {}

    private let timeout: TimeInterval = 10 * 60  // 10 minutos

    /// Sessão expirou por inatividade? A UI observa e reage.
    @Published private(set) var expired = false

    private var armed = false
    private var ultimaInteracao = Date()
    private var timer: Timer?

    /// Liga o vigia — chamar quando o usuário entra numa área logada.
    func arm() {
        armed = true
        expired = false
        ultimaInteracao = Date()
        agendar()
    }

    /// Desliga — logout manual/expiração já tratados.
    func disarm() {
        armed = false
        timer?.invalidate()
        timer = nil
    }

    /// Registra interação do usuário — adia o timeout.
    func touch() {
        guard armed else { return }
        ultimaInteracao = Date()
    }

    /// Chamado quando o app volta do segundo plano: se passou do limite
    /// enquanto estava fora, expira na hora.
    func checarAoVoltar() {
        guard armed else { return }
        if Date().timeIntervalSince(ultimaInteracao) >= timeout { expirar() }
    }

    func consumirExpiracao() { expired = false }

    private func agendar() {
        timer?.invalidate()
        // Checagem a cada 15s — barata e suficiente para um limite de 10 min.
        timer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self, self.armed else { return }
                if Date().timeIntervalSince(self.ultimaInteracao) >= self.timeout {
                    self.expirar()
                }
            }
        }
    }

    private func expirar() {
        disarm()
        expired = true
    }
}

extension View {
    /// Instala um detector de toque que não interfere nos gestos da tela
    /// (o `.simultaneousGesture` deixa botões e scroll funcionarem normalmente).
    func detectaInteracao() -> some View {
        simultaneousGesture(
            DragGesture(minimumDistance: 0, coordinateSpace: .global)
                .onChanged { _ in SessionWatcher.shared.touch() },
            including: .all
        )
    }
}
